### ISSUE TITLE
Abstract class Component and constructor injection.

### ISSUE THREAD

**Issue by rkapsi** [google/dagger#193]
Hello,

I'm not sure if this makes sense or not but I have an use-case for an abstract class Component with an injectable Constructor (and probably for methods and fields). It's currently failing with `@Inject is nonsense on the constructor of an abstract class` error. It makes sense under normal circumstances but what if the Component itself is the concrete "thing" that I'm building. I know I could add a bunch of provision methods to Component but that is potentially exposing more than I want to (and I dislike abstract classes calling abstract methods).

Right now I'm using an interface Component which has a `getMyThing()` provision method and the component gets thrown away once I've retrieved the constructed object.

I could put a wrapper around the Component but then I'd have to do something about the Builder (which I really like about the Components).

```
Dagger 2.1-SNAPSHOT
```

```
@Component(modules=MyConfigModule.class)
@Singleton
public abstract class MyComponent {

  private final MyConfig config;
  
  @Inject
  public MyComponent(MyConfig config) {
    this.config = config;
  }
  
  public void doSomething() {
    System.out.println("Do something: " + config);
  }
  
  @Component.Builder
  public interface Builder {
    
    public Builder config(MyConfigModule module);
    
    public MyComponent build();
  }
  
  public static class MyConfig {}
  
  @Module
  public static class MyConfigModule {
    private final MyConfig config;

    public MyConfigModule(MyConfig config) {
      this.config = config;
    }
    
    public MyConfig provideMyConfig() {
      return config;
    }
  }
  
  public static void main(String[] args) {
    MyComponent component = DaggerMyComponent.builder()
        .config(new MyConfigModule(new MyConfig() {
          @Override
          public String toString() {
            return "Hello, World!";
          }
        }))
        .build();
    
    component.doSomething();
  }
}
```

**Comment by cgruber** [comment]
While I agree that having the thing and the component for the thing be one
thing is feasible, we decided as a matter of policy to not support this,
since the degenerate error cases are confusing enough that this would be
worse than what it's solving.

I tend to do things like this:

class Foo {
  @dagger.Component(modules=...)
  interface Component {
    return Foo();
  }

  @Inject Foo(...) {...}

  ...
}

This way I can reference Foo.Component conveniently, or DaggerFoo_Component
for the impl, and the resulting client code is pretty readable.

Remember - the Component represents the graph, @Inject classes are things
within the graph.  Keeping the types separate maintains that separation
somewhat.  While I acknowledge that you *could* (if we permitted it)
construct something the way you're trying to, the gain isn't worth the loss
of preventing certain classes of error.
